packages/component-base/src/data-provider-controller/data-provider-controller.js

```diff
@@ -0,0 +1,241 @@
+/**
+ * @license
+ * Copyright (c) 2021 - 2023 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+import { Cache } from './cache.js';
+import { getFlatIndexByPath, getFlatIndexContext } from './helpers.js';
+
+/**
+ * A controller that stores and manages items loaded with a data provider.
+ */
+export class DataProviderController extends EventTarget {
+  /**
+   * The controller host element.
+   *
+   * @param {HTMLElement}
+   */
+  host;
+
+  /**
+   * A callback that returns data based on the passed params such as
+   * `page`, `pageSize`, `parentItem`, etc.
+   */
+  dataProvider;
+
+  /**
+   * A callback that returns additional params that need to be passed
+   * to the data provider callback with every request.
+   */
+  dataProviderParams;
+
+  /**
+   * A number of items in the root cache.
+   *
+   * @type {number}
+   */
+  size;
+
+  /**
+   * A number of items to display per page.
+   *
+   * @type {number}
+   */
+  pageSize;
+
+  /**
+   * A callback that returns whether the given item is expanded.
+   *
+   * @type {(item: unknown) => boolean}
+   */
+  isExpanded;
+
+  /**
+   * A reference to the root cache instance.
+   *
+   * @param {Cache}
+   */
+  rootCache;
+
+  constructor(host, { size, pageSize, isExpanded, dataProvider, dataProviderParams }) {
+    super();
+    this.host = host;
+    this.size = size;
+    this.pageSize = pageSize;
+    this.isExpanded = isExpanded;
+    this.dataProvider = dataProvider;
+    this.dataProviderParams = dataProviderParams;
+    this.rootCache = this.__createRootCache();
+  }
+
+  /**
+   * The total number of items, including items from expanded sub-caches.
+   */
+  get effectiveSize() {
+    return this.rootCache.effectiveSize;
+  }
+
+  /** @private */
+  get __cacheContext() {
+    return { isExpanded: this.isExpanded };
+  }
+
+  /**
+   * Whether the root cache or any of its decendant caches have pending requests.
+   *
+   * @return {boolean}
+   */
+  isLoading() {
+    return this.rootCache.isLoading;
+  }
+
+  /**
+   * Sets the size for the root cache and recalculates the effective size.
+   *
+   * @param {number} size
+   */
+  setSize(size) {
+    this.size = size;
+    this.rootCache.size = size;
+    this.recalculateEffectiveSize();
+  }
+
+  /**
+   * Sets the page size and clears the cache.
+   *
+   * @param {number} pageSize
+   */
+  setPageSize(pageSize) {
+    this.pageSize = pageSize;
+    this.clearCache();
+  }
+
+  /**
+   * Sets the data provider callback and clears the cache.
+   *
+   * @type {Function}
+   */
+  setDataProvider(dataProvider) {
+    this.dataProvider = dataProvider;
+    this.clearCache();
+  }
```

Comment:
nit: The tests pass without the `this.clearCache()` calls here. I assume it's because the mixin also invokes clear cache when page size or data provider changes.

Reply:
I guess it works because there is still a `clearCache()` call here:

https://github.com/vaadin/web-components/blob/398e4c342b380ebc1250411c99ab6d10fb2cd370/packages/grid/src/vaadin-grid-data-provider-mixin.js#L350-L354

However, I'd still prefer keeping `clearCache()` in `setDataProvider()`. I believe this makes sense when we view `DataProviderController` as an abstract thing. I'm also planning to cover this with a test later.

Reply:
In one of the early revisions, as an experiment, I partially inlined `grid.clearCache()` into `grid.__dataProviderChanged()` without actually calling `dataProviderController.clearCache()`. Then I decided to revert those changes to keep the diff smaller.